File: src/my_py_pkg/my_py_pkg/number_publisher.py
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64


class NumberPublisher(Node):

    def __init__(self):
        super().__init__("number_publisher_client")
        self.counter_ = 0
        self.publisher_ = self.create_publisher(Int64, "number", 10)
        self.get_logger().info("published on topic number_publisher")
        self.timer_ = self.create_timer(1, self.callback_number_publisher)

    def callback_number_publisher(self):
        self.counter_ += 1
        msg = Int64()
        msg.data = self.counter_
        self.publisher_.publish(msg)
        self.get_logger().info("publishing = " + str(self.counter_))


def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisher()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/number_publisher.cpp
#include "example_interfaces/msg/int64.hpp"
#include "rclcpp/rclcpp.hpp"

class NumberPublisherNode : public rclcpp::Node {
public:
    NumberPublisherNode() : Node("number_publisher_node") {
        counter_ = 0;
        number_publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&NumberPublisherNode::callbackNumberPublisher, this));
    };

private:
    int counter_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr number_publisher_;
    rclcpp::TimerBase::SharedPtr timer_;

    void callbackNumberPublisher() {
        counter_ += 1;
        auto msg = example_interfaces::msg::Int64();
        msg.data = counter_;
        number_publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "published %d", counter_);
    };
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
};